sys/kern/src/arch/arm_m.rs

```diff
@@ -571,30 +576,37 @@ pub fn apply_memory_protection(task: &task::Task) {
         disable_mpu(mpu);
     }
 
+    // We'll collect the MAIR register contents here. Indices 0-3 correspond to
+    // MAIR0's bytes (in LE order); 4-7 are MAIR1.
+    let mut mairs = [0; 8];
+
     for (i, region) in task.region_table().iter().enumerate() {
         let rnr = i as u32;
 
         let ext = &region.arch_data;
 
+        mairs[i] = ext.mair;
+
+        // Set the attridx field of the RLAR to just choose the attributes with
+        // the same index as the region. This lets us treat MAIR as an array
+        // corresponding to the regions.
+        //
+        // We unfortunately can't do this at compile time, because regions can
+        // be shared, and may not be used in the same table position in all
+        // tasks.
+        let rlar = ext.rlar | (i as u32) << 1; // AttrIdx
+
         unsafe {
-            let rlar_disabled = ext.rlar_disabled | (i as u32) << 1; // AttrIdx
             mpu.rnr.write(rnr);
-            mpu.rlar.write(rlar_disabled); // configure but leave disabled
-            if rnr < 4 {
-                let mut mair0 = mpu.mair[0].read();
-                mair0 |= ext.mair << (rnr * 8);
-                mpu.mair[0].write(mair0);
-            } else {
-                let mut mair1 = mpu.mair[1].read();
-                mair1 |= ext.mair << ((rnr - 4) * 8);
-                mpu.mair[1].write(mair1);
-            }
             mpu.rbar.write(ext.rbar);
-            mpu.rlar.write(rlar_disabled | 1); // enable the region
+            mpu.rlar.write(rlar);
         }
     }
 
     unsafe {
+        // Load the MAIR registers.
+        mpu.mair[0].write(u32::from_le_bytes(mairs[..4].try_into().unwrap()));
+        mpu.mair[1].write(u32::from_le_bytes(mairs[4..].try_into().unwrap()));
```

Comment:
hm, the `try_into()` is a bit of a bummer here, since these are always going to be fixed-size arrays, and it would be nice to be able to reason about that statically and avoid checking that they actually are 4 bytes at runtime. but, i guess the alternative is to declare two 4-byte arrays and make the `for` loop above do some extra logic to reason about which one it's in, which also feels kind of sad. i dunno...